src/v/raft/state_machine.cc

```diff
@@ -105,10 +113,16 @@ ss::future<> state_machine::apply() {
     return _raft->events()
       .wait(_next, model::no_timeout, _as)
       .then([this] {
-          // build a reader for log range [_next, +inf).
-          storage::log_reader_config config(
-            _next, model::model_limits<model::offset>::max(), _io_prio);
-          return _raft->make_reader(config);
+          auto f = ss::now();
+          if (_next < _raft->start_offset()) {
+              f = handle_eviction();
+          }
+          return f.then([this] {
```

Comment:
I guess in theory additional eviction can happen between the check and creating the reader, so it would be cleaner to recurse. But if this happens, the end result will be that the exception will be thrown and printed and recursion will happen anyway.

Reply:
nice. thanks for the explanation!